BUG (string dtype): comparison of string column to mixed object column fails #60228 (fixed)

TEARFEAR opened the pull request:
#### Changes :
- Updated `comparison_op` function in `pandas/core/ops/array_ops.py`:
  - Added handling for comparisons between `string` and `object` dtypes.
  - Ensured `string` and `object` types are casted to a common type (`string`) before comparison.
- Added a test case to validate the fix:
  - Verified comparison works for `string` vs `object` with homogeneous and mixed types.
  - Verified behavior with PyArrow-based strings enabled (`pd.options.future.infer_string = True`).

--

- [x] closes #60228  (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

TEARFEAR committed:
fixed comparison of string column to mixed object column (issue #60228)

TEARFEAR committed:
fixed comparison of string column to mixed object column (issue #60228)

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
Merge branch 'main' into bug-update-60228

KevsterAmp committed:
CI/BUG: Remove `trim()` function on `comment-commands.yml` (#60397)

remove trim function on comment-commands.yml

Commit:
DOC: Fixed spelling of 'behaviour' to 'behavior' (#60398)

ZKaoChi committed:
BUG: Convert output type in Excel for MultiIndex with period levels (#60182)

jorisvandenbossche commented on pandas/core/ops/array_ops.py:
```diff
@@ -321,6 +325,16 @@ def comparison_op(left: ArrayLike, right: Any, op) -> ArrayLike:
                 "Lengths must match to compare", lvalues.shape, rvalues.shape
             )
 
+    if (is_string_dtype(lvalues) and is_object_dtype(rvalues)) or (
+        is_object_dtype(lvalues) and is_string_dtype(rvalues)
```
Checking for string dtype for an array _can_ be expensive in case the array is object dtype (at that point it will scan all values to check if they are strings). So we might want to try avoid that at this level. 
I _think_ we could handle the issue specifically for the ArrowExtensionArray itself (see the code I referenced in https://github.com/pandas-dev/pandas/issues/60228#issuecomment-2462533868)

jorisvandenbossche commented on pandas/core/ops/array_ops.py:
```diff
@@ -321,6 +325,16 @@ def comparison_op(left: ArrayLike, right: Any, op) -> ArrayLike:
+    ):
+        if lvalues.dtype.name == "string" and rvalues.dtype == object:
+            lvalues = lvalues.astype("string")
+            rvalues = pd_array(rvalues, dtype="string")
```
We might need to do the casting the other way around. Instead of casting the object to string and then compare both as strings, I think we have to cast the string to object and compare both as object dtype.

The reason for this is that casting to string might actually convert values to strings, and then we are no longer doing the comparison for the original values.

```python
>>> ser_string = pd.Series(["1", "b"])
>>> ser_mixed = pd.Series([1, "b"])
>>> ser_string == ser_mixed
0    False
1     True
dtype: bool

>>> ser_string == ser_mixed.astype("string")
0    True
1    True
dtype: bool
```

So if we would do that casting under the hood, the result would change in this case.

jorisvandenbossche replied:
And we should add this case to the tests!

jorisvandenbossche commented on pandas/tests/series/methods/test_compare.py:
```diff
@@ -138,3 +138,17 @@ def test_compare_datetime64_and_string():
+
+def test_comparison_string_mixed_object():
+    # Issue https://github.com/pandas-dev/pandas/issues/60228
+    pd.options.future.infer_string = True
```
You don't need to add this for CI, because we have a separate CI build that enables this option for the full test suite.

Now, this can still be useful to test locally, but the way you can do this is with setting an environment variable (on linux I can do `PANDAS_FUTURE_INFER_STRING=1 pytest ...` to run the test with the option enabled.

jorisvandenbossche reviewed (commented):
Thanks for working on this! 
Added a few suggestions

partev committed:
fix issue #60410 (#60412)

tuhinsharma121 committed:
DOC: fix SA01 for pandas.errors.UnsortedIndexError (#60404)

lfffkh committed:
Fix BUG: Cannot shift Intervals that are not closed='right' (the default) (#60407)

first

tuhinsharma121 committed:
DOC: fix SA01,ES01 for pandas.errors.PossibleDataLossError (#60403)

tuhinsharma121 committed:
DOC: fix SA01 for pandas.errors.OutOfBoundsTimedelta (#60402)

tuhinsharma121 committed:
DOC: fix SA01,ES01 for pandas.errors.DuplicateLabelError (#60399)

tuhinsharma121 committed:
DOC: fix SA01,ES01 for pandas.errors.InvalidIndexError (#60400)

tuhinsharma121 committed:
DOC: fix SA01 for pandas.errors.NumExprClobberingError (#60401)

mroeschke committed:
TST: Avoid hashing np.timedelta64 without unit (#60416)

snitish committed:
BUG: Fix formatting of complex numbers with exponents (#60417)

Fix formatting of complex numbers with exponents

dependabot[bot] committed:
Bump pypa/cibuildwheel from 2.21.3 to 2.22.0 (#60414)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.21.3 to 2.22.0.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.21.3...v2.22.0)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

sooooooing committed:
DOC: fix docstring api.types.is_re_compilable (#60419)

* fix docstring api.types.is_re_compilable

* fix lint error

ohe committed:
DOC: Clarifying pandas.melt method documentation by replacing "massage" by "reshape" (#60420)

Clarifying pandas.melt method documentation by replacing "massage" by "reshape"

Meanwhile, "massage" is correct in a colloquial sense to mean transforming or reshaping data. This is far from accessible for a non-English speaker (as I am).

Using the term `reshape` or `transform` is more meaningful while being accurate.

partev committed:
replace twitter->X (#60426)

jorisvandenbossche committed:
String dtype: use ObjectEngine for indexing for now correctness over performance (#60329)

jasonmokk committed:
DOC: Add type hint for squeeze method (#60415)

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

yuanx749 committed:
BUG: fix NameError raised when specifying dtype with string having "[pyarrow]" while PyArrow is not installed (#60413)

* Add test

* Fix

* Add note

* Update pandas/tests/dtypes/test_common.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

* update

* Fix doc warning

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

TEARFEAR committed:
Merge remote-tracking branch 'origin/bug-update-60228' into bug-update-60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
Merge remote-tracking branch 'origin/bug-update-60228' into bug-update-60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

TEARFEAR committed:
BUG (string dtype): comparison of string column to mixed object column fails #60228

jorisvandenbossche commented:
@TEARFEAR git workflow related comment: it seems you have picked up some unrelated commits. In general doing something like

```
git fetch upstream
git merge upstream/main
git push origin bug-update-60228
```

should fix that (assuming `origin` is your fork and `upstream` is the pandas-dev/pandas remote)

jorisvandenbossche commented:
@TEARFEAR would you have time to update this PR? (otherwise I am also happy to update it, as we would like to include this in the 2.3 release)

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
